pkg/driver/internal/inflight.go

```diff
@@ -0,0 +1,57 @@
+package internal
+
+import (
+	"sync"
+)
+
+// Idempotent is the interface required to manage in flight requests.
+type Idempotent interface {
+	// The CSI data types are generated using a protobuf. The
+	// generated structures are guaranteed to implement the
+	// Stringer interface. Example
+	// https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L3508
+	// We can use the generated string as the key of our internal
+	// inflight database of requests.
+	String() string
+}
+
+// InFlight is a struct used to manage in flight requests.
+type InFlight struct {
+	mux      *sync.Mutex
+	inFlight map[string]bool
+}
+
+// NewInFlight instanciates a InFlight structures.
+func NewInFlight() *InFlight {
+	return &InFlight{
+		mux:      &sync.Mutex{},
+		inFlight: make(map[string]bool),
+	}
+}
+
+// Upsert inserts the entry to the current list of inflight
+// requests. If it's not then it's added and returns true. Returns
+// false if request was not added to list of inflight requests.
+func (db *InFlight) Upsert(entry Idempotent) bool {
```

Comment:
The function name `Upsert` doesn't seem right here as its not really updating anything. I think `Insert` makes more sense, and change the description to point out that it will return false only when the key already exists.